src/rust/grapl-observability/src/statsd_formatter.rs

```diff
@@ -0,0 +1,161 @@
+use crate::metric_error::MetricError;
+use failure::Error;
+use itertools::join;
+use lazy_static::lazy_static;
+use regex::Regex;
+
+lazy_static! {
+    static ref INVALID_CHARS: Regex = Regex::new("[|#,=:]").unwrap();
+}
+
+// TODO look at lifetimes, should this own or not, etc
+// seems odd to .to_string() everytihng going in here
+pub struct TagPair {
+    tag_key: String,
+    tag_value: String,
+}
+
+impl TagPair {
+    fn statsd_serialized(&self) -> String {
+        return format!("{}={}", self.tag_key, self.tag_value);
+    }
+}
+
+fn reject_invalid_chars(s: &str) -> Result<(), Error> {
+    let matched = INVALID_CHARS.is_match(s);
+    if matched {
+        Err(Error::from(MetricError {}))
+    } else {
+        Ok(())
+    }
+}
+
+pub enum MetricType {
+    Gauge,
+    Counter,
+    Millis,
+    Histogram,
+}
+
+impl MetricType {
+    fn statsd_type(&self) -> String {
```

Comment:
These are all static strings so you can just return a &'static str . That'll avoid the allocation.